### DISCUSSION TITLE
Handle `packageManager` in `package.json`

### DISCUSSION THREAD

**Discussion by viraptor** [jdx/mise#3549] [Ideas]
When using mise in a project with `packageManager` defined in `package.json`, I would expect that package to be installed/used in the same way `.foo-version` files are used.

But this:
```
{
  "name": "foo",
  "version": "1.2.3",
  "packageManager": "yarn@1.22.19"
}
```
and
```
mise x yarn -- yarn
```
ignores the configured version and runs (for example) v4.x.x

**Comment by jdx** [comment]
related: https://github.com/jdx/mise/issues/1863

I think this could be done as a vfox/asdf/core plugin which just uses whatever `node` is on PATH. That wouldn't use corepack but I'm wondering if that ultimately matters to the user.

**Comment by viraptor** [comment]
Isn't the node choice already done in the yarn plugin? I can `mise x yarn@1.2.3 -- yarn install` and it does the "right thing" in this case.

The only remaining issue seems to be not selecting the default version through the json file.

**Comment by jdx** [comment]
that would be a feature request for asdf-yarn

**Comment by NullVoxPopuli** [comment]
why asdf-yarn? I don't use yarn, but need packageManager support

**Comment by jdx** [comment]
ok, then asdf-pnpm

**Reply by alper** [reply]
Is that this repository? https://github.com/jonathanmorley/asdf-pnpm

I know the situation for node is a mess, but looking around the only solid place to get this information seems to be this field (without adding another configuration file). So wouldn't it make sense for `package.json` to be added to the idiomatic filenames for node?

**Comment by NullVoxPopuli** [comment]
ah, no `packageManager` plugin? each separate package manager has to separately add support?

**Comment by jdx** [comment]
it would be possible for someone to make one that installs whichever one is needed but I don't think it's necessary

**Comment by jguddas** [comment]
I have`pnpm = "latest"` in my `~/.config/mise/config.toml`, to use the version of pnpm specified in the `packageManager` field
I have added this to the `mise.toml` that is located besides the `package.json`:
```toml
[hooks]
postinstall = "corepack enable"

[settings]
disable_tools = ["pnpm"]
```

And that works pretty well, now the shimming of pnpm is done by corepack, but only for this project, everything else still uses mise.

**Comment by cprecioso** [comment]
Just wanted to share my workaround, which works fine enough and doesn't depend on postinstall, 

```toml
[tools]
node = "latest"
"npm:corepack" = "latest"
```

It will work perfectly

**Comment by sandros94** [comment]
I'm new to mise and trying to navigate old discussions, unavailable issues and incomplete docs.

So the only way to read from the `package.json` the current `engines.node` and `packageManager` is to also have a `mise.toml` sidecar committed to the project?

There isn't an automatic way to always read the `package.json`, since it is the only agnostic way to get that information?

**Reply by cprecioso** [reply]
The method I mentioned in my previous post works perfectly https://github.com/jdx/mise/discussions/3549#discussioncomment-11918572.
Just run `mise use npm:corepack` and it will pick up your `packageManager` keys.

**Comment by risu729** [comment]
This is now supported with https://github.com/jdx/mise/pull/8059!
